### PR TITLE
[WIP] change pytorch preprocessing code to tensorflow

### PR DESCRIPTION
**Pull request by lsg1213**

### CODE REVIEW

File: data_loader.py [hunk 1]
```diff
@@ -122,12 +124,11 @@ def preprocess_label(labels, max_label_length=600):
...
 
         if cur_len < max_len: 
-            labels = np.pad(labels, ((0, max_len-cur_len), (0,0)), 'constant')
+            labels = tf.pad(labels, ((0, max_len-cur_len), (0,0)))
         else:
             labels = labels[:max_len]
         return labels
-    
-    x = list(map(lambda x: torchaudio.load(x)[0], f_paths))
+    x = list(map(lambda x: tf.audio.decode_wav(tf.io.read_file(x))[0], f_paths))
```
Comment (daniel03c1) [inline data_loader.py:131]: 여기에 아마 transpose를 넣어야 될 것 같습니다. torchaudio랑 모양이 반대입니다.

File: data_loader.py [hunk 2]
```diff
@@ -174,13 +175,12 @@ def get_TDMset(TDM_PATH):
     from glob import glob
     tdm_path = os.path.join(TDM_PATH, 'foa_dev_tdm')
     class_num = len(glob(tdm_path + '/*label_*.joblib'))
-    device = get_device()
 
     def load_data(cls):
-        return joblib.load(os.path.join(tdm_path, f'tdm_noise_{cls}.joblib'))
+        return tf.cast(tf.transpose(tf.convert_to_tensor(joblib.load(os.path.join(tdm_path, f'tdm_noise_{cls}.joblib'))), (1,0)), tf.float32)
```
Comment (daniel03c1) [inline data_loader.py]: convert_to_tensor에서 직접 dtype을 지정하면 cast가 필요없을 것 같습니다.
그리고 transpose할때 어짜피 음성의 ndim는 2이기 때문에 (1, 0)하지 않고 그냥
transpose만 해도 동일한 효과를 얻을 수 있습니다!

File: data_loader.py [hunk 3]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
 
 def TDM_aug(x: list, y: list, tdm_x, tdm_y, sr=24000, label_resolution=0.1, max_overlap_num=5, max_overlap_per_frame=2, min_overlap_time=1, max_overlap_time=5):
```
Comment (daniel03c1) [inline data_loader.py]: min_overlap_time하면 frame단위인지 sec단위인지 헷깔리는데 sec로 바로 표현하는건 어떤가요?

File: data_loader.py [hunk 4]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
-    weight = 1 / torch.tensor([i.shape[0] for i in tdm_y])
-    weight /= weight.sum()
-    weight = weight.cumsum(-1)
+
+    weight = 1 / tf.convert_to_tensor([i.shape[0] for i in tdm_y])
+    weight /= tf.reduce_sum(weight)
+    # weight = tf.math.cumsum(weight)
     def add_noise(i):
-        selected_cls = weight.multinomial(max_overlap_num, replacement=True) # (max_overlap_num,)
+        selected_cls = tf.random.categorical(weight[tf.newaxis,...], max_overlap_num)[0] # (max_overlap_num,)
```
Comment (daniel03c1) [inline data_loader.py]: categorical의 입력으로 logits이 들어가니까, 그에 맞게 weight을 조절해야 할 것 같습니다.

File: data_loader.py (hunk 1 above)
Comment (lsg1213) [reply data_loader.py:131]: 그래서 실제 합성하고 사용할 때에는 tf식 모양에 맞춰서 진행하고 있습니다

File: data_loader.py (hunk 2 above)
Comment (lsg1213) [reply data_loader.py]: 감사합니다!

File: data_loader.py (hunk 3 above)
Comment (lsg1213) [reply data_loader.py]: 넵 수정했습니다

File: data_loader.py (hunk 4 above)
Comment (lsg1213) [reply data_loader.py]: 이 부분이 이해가 좀 안 되는데 [0.2 0.8] 이렇게 넣으면 0이 뽑힐 확률이 0.2고 1이 뽑힐 확률이 0.8이 아닌가요?

File: data_loader.py [hunk 5]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
     sr = int(sr * label_resolution)
-    
-    weight = 1 / torch.tensor([i.shape[0] for i in tdm_y])
-    weight /= weight.sum()
-    weight = weight.cumsum(-1)
+
+    weight = 1 / tf.convert_to_tensor([i.shape[0] for i in tdm_y])
+    weight /= tf.reduce_sum(weight)
+    # weight = tf.math.cumsum(weight)
     def add_noise(i):
```
Comment (daniel03c1) [inline data_loader.py]: add noise 함수가 y 등을 global variable 참조하듯이 진행하는데, 
가능하다면 x, y와 같은 필요한 인자들을 다 받는 함수로 하는게 더 안전하지 않을까 싶습니다.

File: data_loader.py [hunk 6]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
-            y_idx = frame_idx + offset # 합칠 프레임들 전체
-            con = (torch.unsqueeze(y_idx, -1) == valid_index).sum(-1) # valid_index 중 y_idx에 있는 idx만 찾기, masking 방식의 결과
-            if con.sum() == 0: # 만약 넣을 수 없다면 이번에는 노이즈 안 넣음
-                continue
-            
-            idx = torch.where(con > 0)[0]
-            y_idx = y_idx[idx] # 해당되지 않는 부분 삭제, 자리 확정
-            td_offset = torch.randint(td_y.shape[0] - noise_time.item(), (1,)) # 뽑을 노이즈에서의 랜덤 offset
-            td_y_idx = idx + td_offset # 뽑을 노이즈 index
+            frame_y = tf.pad(tf.ones((sample_time, 1)), ((offset, frame_y_num - offset - sample_time), (0,0))) # 샘플과 동일한 크기 생성
```
Comment (daniel03c1) [inline data_loader.py]: tf.ones((sample_time,)) 이런식으로만 해서 padding한 다음에 expand_dims나 [..., tf.newaxis] 형태로 하면
어떨까요?

File: data_loader.py [hunk 7]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
+            frame_y *= valid_index # valid한 프레임만 남기기
+            if tf.reduce_sum(frame_y) == 0: # 만약 넣을 수 없다면 이번에는 노이즈 안 넣음
+                return tf.zeros((), dtype=tf.int64) # dummy return for using tf.map_fn
```
Comment (daniel03c1) [inline data_loader.py]: 아무의미 없는 출력이라면 그냥 return 만 적는건 어떤가요?

File: data_loader.py [hunk 8]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
-            y[i][y_idx] = (ys[y_idx] + td_y[td_y_idx]).cpu() # 레이블 부분 완료
-    
-    # with ThreadPoolExecutor() as pool:
-    #     list(pool.map(add_noise, tqdm(range(len(x)))))
-    # with ProcessPoolExecutor(cpu_count() // 2) as pool:
-    #     list(pool.map(add_noise, tqdm(range(len(x)))))
-    list(map(add_noise, tqdm(range(len(x)))))
-    return x, y
+    norm = tf.math.sqrt(IVx**2 + IVy**2 + IVz**2)
+    norm = tf.math.maximum(norm, tf.zeros_like(norm)+eps)
```
Comment (daniel03c1) [inline data_loader.py]: 여기 tf.math.maximum(norm, eps)로 바꿔도 될 것 같습니다.

File: data_loader.py [hunk 9]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
 
 def TDM_aug(x: list, y: list, tdm_x, tdm_y, sr=24000, label_resolution=0.1, max_overlap_num=5, max_overlap_per_frame=2, min_overlap_time=1, max_overlap_time=5):
     '''
-        x: list(torch.Tensor): shape(sample number, channel(4), frame(1440000))
-        y: list(np.ndarray): shape(sample number, time(600), class+cartesian(14+42))
-        tdm_x: list(np.ndarray): shape(class_num, channel(4), frame)
-        tdm_y: list(np.ndarray): shape(class_num, time, class+cartesian(14+42))
+        x: list(tf.Tensor): shape(sample number, frame(1440000), channel(4))
```
Comment (daniel03c1) [inline data_loader.py:193]: 여기서 sample number가 batch size 말씀이신건가요?

File: data_loader.py (hunk 7 above)
Comment (lsg1213) [reply data_loader.py]: None type return을 지원하지 않는다고 해서 저렇게 했습니다.

File: data_loader.py (hunk 9 above)
Comment (lsg1213) [reply data_loader.py:193]: preprocessing 전이라 train셋에 사용되는 wav 400개를 의미합니다.

File: data_loader.py (hunk 8 above)
Comment (lsg1213) [reply data_loader.py]: 저 부분은 급한대로 진혁씨 코드 긁어온거라 나중에 진혁씨 머지를 기다리는 게 더 좋지 않을까 라는 생각이 들었습니다

File: data_loader.py (hunk 6 above)
Comment (lsg1213) [reply data_loader.py]: 연산량이 더 많지 않나요? expand를 할 때 통째로 tensor를 옮기는 걸로 알고 있는데 제가 잘못알고 있을수도 있고요
그리고 tf.pad는 1d padding을 지원하지 않더라고요

File: data_loader.py [hunk 10]
```diff
@@ -190,60 +190,80 @@ def load_label(cls):
...
-            con = (torch.unsqueeze(y_idx, -1) == valid_index).sum(-1) # valid_index 중 y_idx에 있는 idx만 찾기, masking 방식의 결과
-            if con.sum() == 0: # 만약 넣을 수 없다면 이번에는 노이즈 안 넣음
-                continue
-            
-            idx = torch.where(con > 0)[0]
-            y_idx = y_idx[idx] # 해당되지 않는 부분 삭제, 자리 확정
-            td_offset = torch.randint(td_y.shape[0] - noise_time.item(), (1,)) # 뽑을 노이즈에서의 랜덤 offset
-            td_y_idx = idx + td_offset # 뽑을 노이즈 index
+            frame_y = tf.pad(tf.ones((sample_time, 1)), ((offset, frame_y_num - offset - sample_time), (0,0))) # 샘플과 동일한 크기 생성
+            frame_y *= valid_index # valid한 프레임만 남기기
```
Comment (daniel03c1) [inline data_loader.py]: frame_y가 더할 y frame mask 역할을 하는데, 이후에 y[I], x[I]와 곱해지기 전까지는 1차원만 있어도 되니까,
219번째 줄에 [..., tf.newaxis], 221번째 줄에 tf.pad 뒤에 (0, 0) 없애기 등을 하고
마지막에 222줄 끝나고 한번 tf.newaxis하면 조금 더 간단해지지 않을까 싶습니다.

Comment (lsg1213) [reply data_loader.py]: 앞서 코멘트에 달은 내용인데 tf.pad가 column이 2개 이상인 경우에만 사용이 가능하다는 에러를 뱉어서 저렇게 했습니다

File: data_loader.py (hunk 1 above)
Comment (daniel03c1) [reply data_loader.py:131]: tf, torchaudio 두가지 방식마다 사용할때 데이터 순서를 신경쓰는 것보다
하나로 통일하는게 더 좋지 않을까 생각이 들어서요. 지금도 조금 복잡하다는 생각이 드는데,
실수할 여지가 조금이라도 생기는 걸 줄이면 좋지 않을까 생각이 듭니다.

File: data_loader.py (hunk 6 above)
Comment (daniel03c1) [reply data_loader.py]: expand는 예시고, [..., tf.newaxis]도 그래서 같이 언급했습니다.
제가 이야기하려는건 채널을 굳이 여기서 맞출 필요가 있나 싶어서 언급했습니다.
pad가 1d 지원이 됩니다! [[offset, frame_y_num-offset-sample_time]] 이런식으로 할 수 있습니다!

File: data_loader.py (hunk 7 above)
Comment (daniel03c1) [reply data_loader.py]: 사실 위에도 언급하긴 했는데, add_noise라는 함수 형태가 위험한 형태의 함수이긴 합니다.
함수 내부에서 argument로 받지 않은 외부 변수를 바꾸면 로직상으로 문제 없어보여도 의도하지 않은 결과물이
나오거나 할 수 있어서 최대한 피해야 하는데, 차라리 x, y를 돌려주면 어떨까 싶습니다.
특히나 map_fn이 parallel하게 연산하게 되어 있는데, 원래 map_fn의 원래 용도는 데이터가 있고
데이터를 받아 함수를 통해 바꾸기 위해서 사용하는건데, 여기서는 add_noise가 sequential하게 돌아가야 정상이고,
입력 외의 데이터를 계속 바꿔서, 사실 데이터가 어떤 이상한 데이터가 나와도 이상하지 않은 상태이긴 합니다.
차라리 tf.while_loop를 활용해서 좀더 안전하면서 빠르게 하면 어떨까 생각이 듭니다.
함수도 안전해져서 테스트하기도 좋아지지 않을까 생각이 듭니다. 그러면 return None과 같은 코드도
없어도 될 것 같아서요

File: data_loader.py (hunk 8 above)
Comment (daniel03c1) [reply data_loader.py]: 아 진혁씨 코드는 나중에 승진씨 코드에 합쳐져서 이용될 걸로 들어서 반영해도 괜찮지 않을까 생각이 드는데 @ParkJinHyeock  어떠신가요?

File: data_loader.py (hunk 1 above)
Comment (lsg1213) [reply data_loader.py:131]: 이 부분같은 경우에는 다시 torch로 돌아갈 일이 없기 때문에 tf방식대로 놔두는 게 좋을 것 같다는 생각으로 그대로 진행했습니다 다니엘씨 의견이 그러시니 그러면 transpose해서 진행할게요

File: data_loader.py (hunk 8 above)
Comment (ParkJinHyeock) [reply data_loader.py]: @daniel03c1  네 저도 사실 이 작업할때 쓰려고 미리 만들어둔거라 한번에 적용이 더나아보입니다.

File: data_loader.py (hunk 6 above)
Comment (lsg1213) [reply data_loader.py]: > expand는 예시고, [..., tf.newaxis]도 그래서 같이 언급했습니다.
> 제가 이야기하려는건 채널을 굳이 여기서 맞출 필요가 있나 싶어서 언급했습니다.
> pad가 1d 지원이 됩니다! [[offset, frame_y_num-offset-sample_time]] 이런식으로 할 수 있습니다!

연산량이 좀 많이지긴 하겠지만 어차피 y 자체가 짧아서 티도 안 나긴 하겠네요 그게 가독성이 좋다면 그렇게 바꿀게요

File: data_loader.py (hunk 5 above)
Comment (lsg1213) [reply data_loader.py]: 수정 코드 올렸습니다

File: data_loader.py (hunk 7 above)
Comment (lsg1213) [reply data_loader.py]: tf.while_loop로 로직 변경했습니다

File: data_loader.py (hunk 8 above)
Comment (lsg1213) [reply data_loader.py]: 변경했습니다

File: data_loader.py [hunk 11]
```diff
@@ -1,10 +1,12 @@
 import numpy as np
 import tensorflow as tf
+import tensorflow_probability as tfp
```
Comment (daniel03c1) [inline data_loader.py:3]: 이거 사용 안하는것 같은데 지워도 좋을 것 같습니다.

Comment (lsg1213) [reply data_loader.py:3]: 넵 지웠습니다

File: data_loader.py (hunk 1 above)
Comment (lsg1213) [reply data_loader.py:131]: 아직 적용이 안 되어서 모든 수정 끝난 후 제일 마지막으로 할게요 수정 후 제가 resolve 하겠습니다

File: data_loader.py [hunk 12]
```diff
@@ -174,76 +175,102 @@ def get_TDMset(TDM_PATH):
...
+            tdm_frame_y = tdm_y[cls][td_offset:td_offset+sample_time] * valid_index[...,tf.newaxis] # valid한 프레임만 남기기
+            y[i] += tf.pad(tdm_frame_y * frame_y, ((offset, frame_y_num - offset - sample_time),(0,0))) # 레이블 부분 완료
+
+            tdm_frame_x = tdm_x[cls][td_offset * sr: (td_offset + sample_time) * sr] * tf.repeat(tf.cast(valid_index, dtype=x[i].dtype), sr, axis=0)[...,tf.newaxis]
+            x[i] += tf.pad(tdm_frame_x, ((offset * sr, x[i].shape[0] - (offset + sample_time) * sr),(0,0)))
+            return x, y
+        
+        j = tf.constant(0)
+        cond = lambda i, j, x, y: j < len(selected_cls)
+        body = lambda i, j, x, y: (i, j+1) + _add_noise(i, j, x, y)
```
Comment (daniel03c1) [inline data_loader.py:232]: j가 랜덤하게 되어야 하는데, 여기서는 그냥 최대 5개면 0-4번 클래스 노이즈를 넣는 것 같습니다.
어떤 클래스를 넣을지 함수의 입력으로 넣지 않고, 랜덤하게 내부적으로 골라서 합치는건 어떤가요?
그리고 생각해보니까, 어짜피 tdm 데이터가 각각 다른 클래스이기 때문에 weight을 주지 않아도
랜덤하게 샘플링하면 uniform하게 샘플링 되지 않을까 싶습니다.
물론 제대로 샘플링하고싶다면, 현재 x, y의 분포를 반영해서, 랜덤으로 노이즈를 넣었을 때 평균에 가깝게 되도록
조절할 수도 있을 것 같습니다.

Comment (lsg1213) [reply data_loader.py:232]: 아 weight[j]를 넣어야되는데 j를 넣었네요 감사합니다.

File: data_loader.py [hunk 13]
```diff
@@ -174,76 +175,102 @@ def get_TDMset(TDM_PATH):
...
+        _, _, x, y = tf.while_loop(cond, body, (i, j, x, y))
+        # tf.map_fn(_add_noise, selected_cls)
+        return x, y
+
+    i = tf.constant(0)
+    cond = lambda i, x, y: i < len(x)
+    body = lambda i, x, y: (i+1,) + add_noise(i, x, y)
+    _, x, y = tf.while_loop(cond, body, (i, x, y))
+    # x, y = tf.map_fn(add_noise, tf.range(len(x)))
```
Comment (daniel03c1) [inline data_loader.py:241]: 제가 오해가 있었네요. add_noise의 경우에 sequential하게 되어야 하지만,
x, y 하나씩 돌아가면서 하는거는 서로 다른 데이터를 건드리기 때문에 map_fn 등을 이용해도
무방할 것 같긴 하네요.

Comment (lsg1213) [reply data_loader.py:241]: 옙 나중에 필요에 의해 복원해야 할수도 있겠는데 큰 부담없이 복원해도 되겠군요 ㅎㅎ

File: train.py [hunk 14]
```diff
@@ -281,11 +282,34 @@ def main(config):
     metric_class = SELDMetrics(
         doa_threshold=config.lad_doa_thresh)
 
+
+    # overlap_num = 1
+    # overlap_time = 1
+    overlap_num = 5
+    overlap_time = 5
```
Comment (ParkJinHyeock) [inline train.py:289]: 저 혹시 이부분을 5로 세팅해두신 이유가있을까요?
